### PR TITLE
Canonicalize FSST arrays without branching

### PR DESCRIPTION
**Pull request by a10y**
FSST saves the compressed bytes into a giant packed string heap (which is stored in a `VarBinArray`).

Canonicalizing currently requires decompressing each string one-at-a-time, so that we can get the uncompressed length to construct the new canonical `VarBinAray`. It turns out that `VarBinArray::bytes_at()` is incredibly expensive when it's backed by an `ArrayView`.

This PR instead

* Decompresses the entire packed heap in a single call to the decompressor. This lets us spend all of our time in the hot decompress loop without expensive branches
* Stores the uncompressed lengths for each string. When canonicalizing, we can turn this into an offsets array. Later when we are using `VarBinView`, we can use the lengths directly to build our `BinaryView` values.

## TODO

- [x] Fix merge conflicts from the `vortex_expect` PR
- [x] Cascading-compress the `uncompressed_offsets`

### CODE REVIEW

**Comment by robert3005** [comment]
Is bytes_at always slow or is it metadata deserialisation for arrayview, or is it something else? If it’s something about arrayviews we should investigate

**Comment by a10y** [comment]
## Before

image (1)

## After

Screenshot 2024-09-10 at 11 17 35 PM

**Comment by a10y** [comment]
@robert3005 `bytes_at` spends all of its time in

* ArrayView::child, which itself seems to spend lots of time in `cumulative_nbuffers`
* Deserializing and validating metadata

I think this PR is the right thing to do regardless of performance issues in `bytes_at`, but I can file a ticket for tracking `bytes_at`.

FWIW this PR basically replicates what BtrBlocks does, except we preserve full random-access by storing both the compressed and uncompressed lengths.

image

**Comment by a10y** [comment]
[From_Clipboard.speedscope-4.json](https://github.com/user-attachments/files/16957507/From_Clipboard.speedscope-4.json)

^ speedscope export from before this PR

**Comment by robert3005** [comment]
@a10y yes I didn't mean that this pr is unnecessary. I was trying to figure out if there's something we can improve about arrayviews. I made https://github.com/spiraldb/vortex/issues/786 to maybe investigate if it's worth optimizing

File: encodings/fsst/src/compress.rs [hunk 1]
```diff
@@ -113,7 +118,9 @@ where
     let symbol_lengths_vec: Vec<u8> = compressor.symbol_lengths().to_vec();
     let symbol_lengths =
         PrimitiveArray::from_vec(symbol_lengths_vec, Validity::NonNullable).into_array();
```
Comment (robert3005) [inline encodings/fsst/src/compress.rs:120]: it's shorter to create non nullable PrimitiveArrays via PrimitiveArray::from(Vec)

File: encodings/fsst/src/canonical.rs [hunk 2]
```diff
@@ -1,56 +1,46 @@
...
+            for len in uncompressed_lens_slice.iter() {
+                offset += len;
+                offsets.push(offset);
             }
 
-            let arrow_array = builder.finish();
+            let offsets_array =
+                PrimitiveArray::from_vec(offsets, Validity::NonNullable).into_array();
+            let uncompressed_bytes_array =
+                PrimitiveArray::from_vec(uncompressed_bytes, Validity::NonNullable).into_array();
```
Comment (robert3005) [inline encodings/fsst/src/canonical.rs]: Just `PrimitiveArray::from`

File: encodings/fsst/src/canonical.rs [hunk 3]
```diff
@@ -1,56 +1,46 @@
...
+            let uncompressed_lens_array = self
+                .uncompressed_lengths()
+                .into_canonical()?
+                .into_primitive()?;
+            let uncompressed_lens_slice = uncompressed_lens_array.maybe_null_slice::<i32>();
+
+            for len in uncompressed_lens_slice.iter() {
+                offset += len;
+                offsets.push(offset);
             }
```
Comment (robert3005) [inline encodings/fsst/src/canonical.rs:40]: Isn't this a scan? I have a variant of this in ChunkedArray constructor

**Review by robert3005 — APPROVED** [review]
I made some style comments

**Review by AdamGS — APPROVED** [review]
LGTM

File: encodings/fsst/src/canonical.rs (hunk 3 above)
Comment (a10y) [reply encodings/fsst/src/canonical.rs:40]: I found the for loop easier to read personally so I just left this as-is

File: encodings/fsst/src/array.rs [hunk 4]
```diff
@@ -54,6 +57,14 @@ impl FSSTArray {
             vortex_bail!(InvalidArgument: "symbols and symbol_lengths arrays must have same length");
         }
 
+        if uncompressed_lengths.len() != codes.len() {
+            vortex_bail!(InvalidArgument: "uncompressed_lengths must be same len as codes");
+        }
+
+        if !uncompressed_lengths.dtype().is_int() {
```
Comment (lwwmanning) [inline encodings/fsst/src/array.rs:64]: extremely minor, but might be worth asserting min >= 0 or that it's unsigned

Comment (a10y) [reply encodings/fsst/src/array.rs:64]: I'll add that in flup

**Pull request merged**